pkg/assets/embedded.go

```diff
@@ -0,0 +1,59 @@
+package assets
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/tilt-dev/tilt/pkg/logger"
+)
+
+//go:embed build
+var build embed.FS
+
+type embeddedServer struct {
+	http.Handler
+}
+
+func NewEmbeddedServer() (Server, error) {
+	var assets fs.FS
+	index, err := build.ReadFile("build/index.html")
+	if err == nil {
+		assets, err = fs.Sub(build, "build")
+	}
+
+	if err != nil {
+		return embeddedServer{}, fmt.Errorf("embedded assets unavailable: %w", err)
+	}
+
+	return embeddedServer{
+		Handler: handleRoutingURLs(http.FileServer(http.FS(assets)), index),
+	}, nil
+}
+
+func (s embeddedServer) Serve(ctx context.Context) error {
+	logger.Get(ctx).Verbosef("Serving embedded Tilt production web assets")
+	<-ctx.Done()
+	return nil
+}
+
+func (s embeddedServer) TearDown(ctx context.Context) {
+}
+
+func handleRoutingURLs(handler http.Handler, index []byte) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		if strings.HasPrefix(req.URL.Path, "/r/") {
```

Comment:
hmm...

the prod and old precompiled servers redirect everything to index.html, and whitelist the set of urls that are supposed to be served from the filesystem. 

your new server does the opposite - it serves everything from the embed FS, and whitelists the set of urls that should redirect to index.html. 

is there a reason you decided to flip the decision logic? seems error prone to have two very different decision sets of decision logic and likely to lead to problems.

Reply:
That wasn't intentional. I can change this to wrap the response from the file server and on 404 serve up index.html. That's basically what you're saying right?